print-vsock.c

```diff
@@ -0,0 +1,243 @@
+	{AF_VSOCK_TRANSPORT_UNKNOWN, "UNKNOWN"},
+	{AF_VSOCK_TRANSPORT_NO_INFO, "NO_INFO"},
+	{AF_VSOCK_TRANSPORT_VIRTIO, "VIRTIO"},
+	{ 0, NULL }
+};
+
+enum af_vsockmon_op {
+	AF_VSOCK_OP_UNKNOWN = 0,
+	AF_VSOCK_OP_CONNECT = 1,
+	AF_VSOCK_OP_DISCONNECT = 2,
+	AF_VSOCK_OP_CONTROL = 3,
+	AF_VSOCK_OP_PAYLOAD = 4,
+};
+
+static const struct tok vsock_op[] = {
+	{AF_VSOCK_OP_UNKNOWN, "UNKNOWN"},
+	{AF_VSOCK_OP_CONNECT, "CONNECT"},
+	{AF_VSOCK_OP_DISCONNECT, "DISCONNECT"},
+	{AF_VSOCK_OP_CONTROL, "CONTROL"},
+	{AF_VSOCK_OP_PAYLOAD, "PAYLOAD"},
+	{ 0, NULL }
+};
+
+enum virtio_vsock_type {
+	VIRTIO_VSOCK_TYPE_STREAM = 1,
+};
+
+static const struct tok virtio_type[] = {
+	{VIRTIO_VSOCK_TYPE_STREAM, "STREAM"},
+	{ 0, NULL }
+};
+
+enum virtio_vsock_op {
+	VIRTIO_VSOCK_OP_INVALID = 0,
+	VIRTIO_VSOCK_OP_REQUEST = 1,
+	VIRTIO_VSOCK_OP_RESPONSE = 2,
+	VIRTIO_VSOCK_OP_RST = 3,
+	VIRTIO_VSOCK_OP_SHUTDOWN = 4,
+	VIRTIO_VSOCK_OP_RW = 5,
+	VIRTIO_VSOCK_OP_CREDIT_UPDATE = 6,
+	VIRTIO_VSOCK_OP_CREDIT_REQUEST = 7,
+};
+
+static const struct tok virtio_op[] = {
+	{VIRTIO_VSOCK_OP_INVALID, "INVALID"},
+	{VIRTIO_VSOCK_OP_REQUEST, "REQUEST"},
+	{VIRTIO_VSOCK_OP_RESPONSE, "RESPONSE"},
+	{VIRTIO_VSOCK_OP_RST, "RST"},
+	{VIRTIO_VSOCK_OP_SHUTDOWN, "SHUTDOWN"},
+	{VIRTIO_VSOCK_OP_RW, "RW"},
+	{VIRTIO_VSOCK_OP_CREDIT_UPDATE, "CREDIT UPDATE"},
+	{VIRTIO_VSOCK_OP_CREDIT_REQUEST, "CREDIT REQUEST"},
+	{ 0, NULL }
+};
+
+/* All fields are little-endian */
+
+struct virtio_vsock_hdr {
+	uint64_t	src_cid;
+	uint64_t	dst_cid;
+	uint32_t	src_port;
+	uint32_t	dst_port;
+	uint32_t	len;
+	uint16_t	type;		/* enum virtio_vsock_type */
+	uint16_t	op;		/* enum virtio_vsock_op */
+	uint32_t	flags;
+	uint32_t	buf_alloc;
+	uint32_t	fwd_cnt;
+} UNALIGNED;
+
+struct af_vsockmon_hdr {
+	uint64_t src_cid;
+	uint64_t dst_cid;
+	uint32_t src_port;
+	uint32_t dst_port;
+	uint16_t op;		/* enum af_vsockmon_op */
+	uint16_t transport;	/* enum af_vosckmon_transport */
+	uint16_t len;		/* size of transport header */
+	uint8_t reserved[2];
+} UNALIGNED;
+
+static void
+vsock_virtio_hdr_print(netdissect_options *ndo, const struct virtio_vsock_hdr *hdr)
+{
+	uint16_t u16_v;
+	uint32_t u32_v;
+
+	u32_v = EXTRACT_LE_32BITS(&hdr->len);
+	ND_PRINT((ndo, "len %u", u32_v));
+
+	u16_v = EXTRACT_LE_16BITS(&hdr->type);
+	ND_PRINT((ndo, ", type %s",
+		  tok2str(virtio_type, "Invalid type (%hu)", u16_v)));
+
+	u16_v = EXTRACT_LE_16BITS(&hdr->op);
+	ND_PRINT((ndo, ", op %s",
+		  tok2str(virtio_op, "Invalid op (%hu)", u16_v)));
+
+	u32_v = EXTRACT_LE_32BITS(&hdr->flags);
+	ND_PRINT((ndo, ", flags %x", u32_v));
+
+	u32_v = EXTRACT_LE_32BITS(&hdr->buf_alloc);
+	ND_PRINT((ndo, ", buf_alloc %u", u32_v));
+
+	u32_v = EXTRACT_LE_32BITS(&hdr->fwd_cnt);
+	ND_PRINT((ndo, ", fwd_cnt %u", u32_v));
+}
+
+static size_t
+vsock_transport_hdr_size(uint16_t transport)
+{
+	switch (transport) {
+		case AF_VSOCK_TRANSPORT_VIRTIO:
+			return sizeof(struct virtio_vsock_hdr);
+		default:
+			return 0;
+	}
+}
+
+static void
```

Comment:
This should probably return a "truncated/not truncated" indication, so that if

    len < sizeof(struct af_vsockmon_hdr) + transport_size

it returns "truncated", otherwise it returns "not truncated", and its caller should check the return value and print the truncated message (`tstr`) if it returns a "truncated" indication.